File: src/mvmctl/api/inputs/_ssh_input.py
"""SSH input resolution — Input → Request → ResolvedSSHInput."""

from __future__ import annotations

import logging
from dataclasses import dataclass
from pathlib import Path

from mvmctl.core._shared import Database
from mvmctl.core.config._service import SettingsService
from mvmctl.core.key._repository import KeyRepository
from mvmctl.core.key._resolver import KeyResolver
from mvmctl.core.key._service import KeyService
from mvmctl.core.vm._repository import VMRepository
from mvmctl.core.vm._resolver import VMResolver
from mvmctl.exceptions import SSHError
from mvmctl.models import VMInstanceItem

logger = logging.getLogger(__name__)


@dataclass
class SSHInput:
    """Raw SSH identifiers from CLI."""

    identifier: str
    user: str | None = None
    key: Path | None = None
    cmd: str | None = None


@dataclass(frozen=True)
class ResolvedSSHInput:
    """Fully resolved SSH connection parameters."""

    target_ip: str
    user: str
    key: Path | None
    cmd: str | None


class SSHRequest:
    """Resolve SSHInput against the database."""

    def __init__(self, inputs: SSHInput, db: Database) -> None:
        self._inputs = inputs
        self._db = db
        self._vm: VMInstanceItem | None = None

    def resolve(self) -> ResolvedSSHInput:
        """Resolve all inputs to explicit values."""
        target_ip = self._resolve_target()
        user = self._resolve_user()
        key = self._resolve_key()
        return ResolvedSSHInput(
            target_ip=target_ip,
            user=user,
            key=key,
            cmd=self._inputs.cmd,
        )

    def _resolve_target(self) -> str:
        """
        Resolve target to an IP address.

        Resolution order:
        1. Use identifier from SSHInput (handles name, IP, MAC, ID prefix)
        2. Try to resolve it as a VM entity
        3. If resolved, returns the VM's ipv4 and sets self._vm for key resolution
        4. If not resolved (e.g. raw IP not in DB), falls back to the raw identifier
        5. Error if no identifier provided at all
        """
        target = self._inputs.identifier

        if target is None:
            raise SSHError(
                "Provide a VM identifier (name, ID prefix, IP, or MAC address)"
            )

        # Try to resolve as a VM entity — handles names, IPs, MACs, ID prefixes
        repo = VMRepository(self._db)
        resolver = VMResolver(repo)
        try:
            self._vm = resolver.resolve(target)
            if self._vm.ipv4:
                return self._vm.ipv4
        except Exception:
            pass

        # Fallback: use the raw identifier (e.g. IP for a VM not in DB)
        return target

    def _resolve_user(self) -> str:
        if self._inputs.user is not None:
            return self._inputs.user
        # Check VM's stored ssh_user
        if self._vm and self._vm.ssh_user:
            return self._vm.ssh_user
        return str(SettingsService.resolve(self._db, "defaults.vm", "ssh_user"))

    def _resolve_key(self) -> Path | None:
        """
        Resolve SSH private key path via the key domain.

        Resolution order:
        1. If --key provided:
           - Try as key name via KeyResolver
           - Try as filesystem path via KeyService validation
        2. If not provided:
           - Check VM's stored ssh_keys (most specific — exact keys injected)
           - Fall back to default keys
        """
        key_repo = KeyRepository(self._db)
        key_resolver = KeyResolver(key_repo)

        if self._inputs.key is not None:
            key_str = str(self._inputs.key)

            # 1a. Try as registered key name via KeyResolver
            try:
                key_item = key_resolver.resolve(key_str)
                if key_item.private_key_path:
                    path = Path(key_item.private_key_path)
                    if path.exists():
                        return path
            except Exception:
                pass

            # 1b. Try as direct filesystem path — validate via KeyService
            path = Path(key_str)
            if path.exists() and path.is_file():
                key_service = KeyService(key_repo)
                content = path.read_text()
                if key_service._is_private_key(content):
                    return path

            raise SSHError(
                f"Key '{key_str}' not found or is not a valid private key"
            )

        # 2. No key provided — check VM's stored ssh_keys (most specific)
        if self._vm and self._vm.ssh_keys:
            for key_id in self._vm.ssh_keys:
                try:
                    key_item = key_resolver.by_id(key_id)
                    if key_item.private_key_path:
                        path = Path(key_item.private_key_path)
                        if path.exists():
                            return path
                except Exception:
                    continue

        # 3. Fall back to default keys
        defaults = key_resolver.get_defaults()
        for key_item in defaults:
            if key_item.private_key_path:
                path = Path(key_item.private_key_path)
                if path.exists():
                    return path

        return None

File: pkg/api/inputs/ssh_input.go
package inputs

import (
	"context"
	"fmt"
	"mvmctl/internal/core/config"
	"mvmctl/internal/core/key"
	"mvmctl/internal/core/vm"
	"mvmctl/internal/lib/model"
	"mvmctl/internal/lib/validators"
	"mvmctl/pkg/errs"
	"os"
)

// SSHInput specifies SSH input.
type SSHInput struct {
	Identifier string            `json:"target"            yaml:"target"`
	User       *string           `json:"user,omitempty"    yaml:"user,omitempty"`
	Key        *string           `json:"key,omitempty"     yaml:"key,omitempty"`
	Cmd        *string           `json:"cmd,omitempty"     yaml:"cmd,omitempty"`
	Timeout    *int              `json:"timeout,omitempty" yaml:"timeout,omitempty"`
	Env        map[string]string `json:"env,omitempty"     yaml:"env,omitempty"`
}

// ResolvedSSHInput specifies resolved SSH input.
type ResolvedSSHInput struct {
	TargetIP string
	User     string
	Key      *string
	Cmd      *string
	Timeout  *int
}

// Validate checks that the SSH input has a target identifier.
func (i *SSHInput) Validate() error {
	if i.Identifier == "" {
		return errs.New(
			errs.CodeSSHError,
			"Provide a VM identifier (name, ID prefix, IP, or MAC address)",
			errs.WithClass(errs.ClassValidation),
		)
	}
	return nil
}

// Resolve resolves the SSH input against the database and config.
// Returns a fully resolved SSH input with target IP, user, and key path.
func (i *SSHInput) Resolve(
	ctx context.Context,
	cfg *config.Service,
	vmRepo vm.Repository,
	keyRepo key.Repository,
) (*ResolvedSSHInput, error) {
	if err := i.Validate(); err != nil {
		return nil, err
	}
	targetIP, vmEntity, err := resolveTarget(ctx, i.Identifier, vmRepo)
	if err != nil {
		return nil, err
	}
	user, err := resolveUser(ctx, i.User, vmEntity, cfg)
	if err != nil {
		return nil, err
	}
	sshKey, err := resolveSSHKey(ctx, i.Key, vmEntity, keyRepo)
	if err != nil {
		return nil, err
	}
	// Validate resolved values.
	if !validators.IsIPAddress(targetIP) {
		return nil, errs.New(
			errs.CodeSSHError,
			fmt.Sprintf("Invalid IP address: %s", targetIP),
			errs.WithClass(errs.ClassValidation),
		)
	}
	if err := validators.SSHUsername(user); err != nil {
		return nil, errs.New(errs.CodeSSHError, err.Error(), errs.WithClass(errs.ClassValidation))
	}
	if sshKey != nil && *sshKey != "" {
		if _, err := os.Stat(*sshKey); os.IsNotExist(err) {
			return nil, errs.NotFound(errs.CodeKeyNotFound, fmt.Sprintf("SSH key not found: %s", *sshKey))
		}
	}
	return &ResolvedSSHInput{
		TargetIP: targetIP,
		User:     user,
		Key:      sshKey,
		Cmd:      i.Cmd,
		Timeout:  i.Timeout,
	}, nil
}

// resolveTarget resolves the target to an IP address and optionally a VM entity.
// Tries VM resolver first, falls back to raw identifier (e.g., IP for a VM not in DB).
func resolveTarget(ctx context.Context, identifier string, vmRepo vm.Repository) (string, *model.VMItem, error) {
	vmResolver := vm.NewResolver(vmRepo)
	vmEntity, err := vmResolver.Resolve(ctx, identifier)
	if err == nil && vmEntity != nil && vmEntity.IPv4 != "" {
		return vmEntity.IPv4, vmEntity, nil
	}
	// Fallback: use raw identifier
	return identifier, nil, nil
}

// resolveUser resolves the SSH user from input, VM, or config default.
func resolveUser(ctx context.Context, inputUser *string, vmEntity *model.VMItem, cfg *config.Service) (string, error) {
	if inputUser != nil && *inputUser != "" {
		return *inputUser, nil
	}
	if vmEntity != nil && vmEntity.SSHUser != nil && *vmEntity.SSHUser != "" {
		return *vmEntity.SSHUser, nil
	}
	user, _ := cfg.GetString(ctx, "defaults.vm", "ssh_user")
	return user, nil
}

// resolveSSHKey resolves SSH private key path via the key domain.
func resolveSSHKey(
	ctx context.Context,
	inputKey *string,
	vmEntity *model.VMItem,
	keyRepo key.Repository,
) (*string, error) {
	keyResolver := key.NewResolver(keyRepo)
	if inputKey != nil && *inputKey != "" {
		keyStr := *inputKey
		// 1a. Try as registered key name via key resolver
		keyItem, err := keyResolver.Resolve(ctx, keyStr)
		if err == nil && keyItem.PrivateKeyPath != nil && *keyItem.PrivateKeyPath != "" {
			if _, err := os.Stat(*keyItem.PrivateKeyPath); err == nil {
				return keyItem.PrivateKeyPath, nil
			}
		}
		// 1b. Try as direct filesystem path — validate private key content
		if fi, err := os.Stat(keyStr); err == nil && !fi.IsDir() {
			content, err := os.ReadFile(keyStr)
			if err == nil && key.IsPrivateKey(string(content)) {
				return &keyStr, nil
			}
		}
		return nil, errs.New(
			errs.CodeSSHError,
			fmt.Sprintf("Key '%s' not found or is not a valid private key", keyStr),
			errs.WithClass(errs.ClassValidation),
		)
	}
	// 2. No key provided — check VM's stored ssh_keys
	// SSHKeys stores key NAMES; use the key resolver which handles
	// resolution by name, ID, or .pub file path.
	if vmEntity != nil {
		for _, keyName := range vmEntity.SSHKeys {
			keyItem, err := keyResolver.Resolve(ctx, keyName)
			if err == nil && keyItem.PrivateKeyPath != nil && *keyItem.PrivateKeyPath != "" {
				if _, err := os.Stat(*keyItem.PrivateKeyPath); err == nil {
					return keyItem.PrivateKeyPath, nil
				}
			}
		}
	}
	// 3. Fall back to default keys
	defaults, err := keyRepo.GetDefaults(ctx)
	if err == nil {
		for _, keyItem := range defaults {
			if keyItem.PrivateKeyPath != nil && *keyItem.PrivateKeyPath != "" {
				if _, err := os.Stat(*keyItem.PrivateKeyPath); err == nil {
					return keyItem.PrivateKeyPath, nil
				}
			}
		}
	}
	return nil, nil
}
